modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/HashJoinNode.java

```diff
@@ -0,0 +1,718 @@
+                                break;
+                            }
+                        }
+
+                        left = null;
+                    }
+                } finally {
+                    inLoop = false;
+                }
+            }
+
+            getMoreOrEnd();
+        }
+    }
+
+    private static class AntiHashJoin<RowT> extends HashJoinNode<RowT> {
+        private AntiHashJoin(
+                ExecutionContext<RowT> ctx,
+                JoinInfo joinInfo,
+                RelDataType leftRowType,
+                RelDataType rightRowType
+        ) {
+            super(ctx, joinInfo, false, leftRowType, rightRowType);
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        protected void join() throws Exception {
+            if (waitingRight == NOT_WAITING) {
+                inLoop = true;
+                try {
+                    while (requested > 0 && (left != null || !leftInBuf.isEmpty())) {
+                        checkState();
+
+                        left = leftInBuf.remove();
+
+                        Collection<RowT> rightRows = lookup(left, touchResults, this);
+
+                        if (rightRows.isEmpty()) {
+                            requested--;
+
+                            downstream().push(left);
+
+                            if (requested == 0) {
+                                break;
+                            }
+                        }
+
+                        left = null;
+                    }
+                } finally {
+                    inLoop = false;
+                }
+            }
+
+            getMoreOrEnd();
+        }
+    }
+
+    private static <RowT> Collection<RowT> lookup(
+            RowT row,
+            boolean processTouched,
+            HashJoinNode<RowT> node
+    ) {
+        Collection<RowT> coll = Collections.emptyList();
+
+        RowFactory<RowT> leftRowFactory = node.handler.factory(node.leftJoinRelatedRowSchema);
+        RowBuilder<RowT> rowBuilder = leftRowFactory.rowBuilder();
+
+        for (Integer entry : node.leftJoinPositions) {
+            Object ent = node.handler.get(entry, row);
+
+            if (ent == null) {
+                return Collections.emptyList();
+            }
+
+            rowBuilder.addField(ent);
+        }
+
+        RowWrapper<RowT> row0 = new RowWrapper<>(rowBuilder.buildAndReset(), node.handler, node.leftJoinPositions.size());
+
+        TouchedCollection<RowT> found = node.hashStore.get(row0);
+
+        if (found != null) {
+            coll = found.items();
+
+            if (processTouched) {
+                found.touched = true;
+            }
+        }
+
+        return coll;
+    }
+
+    private static <RowT> Iterator<RowT> getUntouched(Map<RowWrapper<RowT>, TouchedCollection<RowT>> entries) {
+        return new Iterator<RowT>() {
+            private final Iterator<Entry<RowWrapper<RowT>, TouchedCollection<RowT>>> outerIt = entries.entrySet().iterator();
```

Comment:
why do we need an entire `Entry`?

Reply:
misunderstand here, i need whole <key, value> record here, or i miss your point ?